refactor(core): add GraphEdge interface to signal graph

leonelvsc opened the pull request:
The idea is to avoid creating extra arrays for the same graph edge info, instead of creating 1 array for each piece of information i refactored to a new interface representing the graph edge which is an array.


## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

alxhub commented:
Hi @leonelvsc,

Consolidating the arrays forces an object allocation for each edge in the graph, which is something we intentionally avoided in the design as it has clear performance cost in benchmarks. Can you elaborate on the rationale for the change?

leonelvsc commented:
> Hi @leonelvsc,
> 
> Consolidating the arrays forces an object allocation for each edge in the graph, which is something we intentionally avoided in the design as it has clear performance cost in benchmarks. Can you elaborate on the rationale for the change?

Hi @alxhub 

I didn't know it was on purpose, my bad. I just saw a lot of unnecessary array ops between index access and array truncating i thought that the cost of allocating an object was more performant in terms of accessing the edges as you just access the array once per operation and more clear and extensible in the future. But don't take this pull request into account it is as always a suggestion from my analysis. 
I think that the main error here from my part is that a plain object was going to add a minimal overhead with those 3 properties in contrast of avoiding extra arrays ops and access.

leonelvsc commented:
Changed the implementation from interface to an array of 3 elements as object access is slower.

atscott commented:
> Changed the implementation from interface to an array of 3 elements as object access is slower.

Could you structure this more like `LView` and `LContainer` (as examples) with the indexes defined as documentable constants?

leonelvsc commented:
> > Changed the implementation from interface to an array of 3 elements as object access is slower.
> 
> Could you structure this more like `LView` and `LContainer` (as examples) with the indexes defined as documentable constants?

Done, please note my english is pretty basic, so feel free to modify in anyway you see necessary the docs or the code itself.

JeanMeche commented on packages/core/primitives/signals/src/graph.ts:
```diff
@@ -298,10 +293,12 @@ export function producerNotifyConsumers(node: ReactiveNode): void {
-    for (const consumer of node.liveConsumerNode) {
-      if (!consumer.dirty) {
-        consumerMarkDirty(consumer);
+    for (let i = 0; i < node.liveConsumerNode.length; i++) {
```
nit: I think we can revert to the initial `for of`, we don't need the index.

leonelvsc replied:
Well i've changed it because this loop is more performant and has 1 alloc which is the index `i,` but feel free to revert it by force-pushing

leonelvsc replied:
@JeanMeche also note, i tried to refactor all with least allocations of memory as possible and passing parameters by reference instead of value, at least is the idea merged or not, i wanted to keep things as fast as possible

leonelvsc commented:
Changed the golden files because some tests were failing, if i changed something wrong please feel free to force push on my branch. Is my first time using the angular dev tools so it's more likely that i've done something wrong.

leonelvsc committed:
refactor(core): add GraphEdge interface to signal graph

The idea is to avoid creating extra arrays for the same graph edge info, instead of creating 1 array for each piece of information i refactored to a new interface representing the graph edge which is an array.

alxhub commented:
@leonelvsc I think this refactor would actually be a regression in performance.

A key property of the existing algorithm is that we avoid allocation when adding/removing edges. This isn't easy to see, as we're `.push()`ing to 3 arrays when adding an edge. But thanks to dynamic array resizing, `.push()` in JavaScript is O(1). The same goes for truncating the array - we only pay the cost of `pop()` and there's no interaction with the GC.

With this change, we have an object which represents the edge, which must be allocated and deallocated as edges are added/removed from the graph. We've seen in benchmarks that the cost of this allocation is non-trivial in a large and dynamic graph.

leonelvsc commented:
> @leonelvsc I think this refactor would actually be a regression in performance.
> 
> A key property of the existing algorithm is that we avoid allocation when adding/removing edges. This isn't easy to see, as we're `.push()`ing to 3 arrays when adding an edge. But thanks to dynamic array resizing, `.push()` in JavaScript is O(1). The same goes for truncating the array - we only pay the cost of `pop()` and there's no interaction with the GC.
> 
> With this change, we have an object which represents the edge, which must be allocated and deallocated as edges are added/removed from the graph. We've seen in benchmarks that the cost of this allocation is non-trivial in a large and dynamic graph.

@alxhub 
First of all thanks i'm really learning a lot about memory management from this. Take into account that passing values instead of references as function parameters makes the engine to alloc memory for each parameter copy, but if you have benchmarked it there's no more discussion about this.

 Well i'm going to close this pull request, just thanks

alxhub commented:
Yeah, these things are fascinating sometimes.

Looking at your jsperf, you're only pushing to the array (it's append-only). This isn't the use case for signals, which have a dynamic but finite set of dependencies. The memory allocation cost is relatively more significant in this pattern.

leonelvsc commented:
@alxhub 
If you don't mind my asking, how about squashing on a single typed array numeric values like `indexOfThis | version` ie: ArrayBuffer or Uint32Array, just for gathering knowallage for those allocs of memory are more performant or same as 52 bit normal integer values "maxInt"?. I'm not saying that's what needs to be done here, just asking for better knowallage about performance and memory.
